### ISSUE TITLE
Can not avoid notifies being shared between machine resources in separate blocks

### ISSUE THREAD

**Issue by poliva83** [chef/chef-provisioning#421]
I have use case where I need to run machine `:destroy` action before `:converge` action but not trigger the same notifies on the `:converge`. Unfortunately the resource name is used to lookup the machine_spec so it appears notifies will be shared even when separating the machine actions into separate DSL blocks. 

Here is simple example recipe (not my real business case) where `destroy_bootstrap_machine` block triggers notifies in `converge_bootstrap_machine` block. My workaround to problem is adding guard on machine_file block to prevent it from attempting :download action when vm is not running.

```
machine "destroy_bootstrap_machine" do
  name bootstrap_machine
  action :destroy
end

machine "converge_bootstrap_machine" do
  name bootstrap_machine
  tag ''
  machine_options :bootstrap_options => { :template => bootstrap_machine_template }
  chef_environment "#{bootstrap_environment}"
  run_list [ 'recipe[bb_apt::default]' ]
  action :converge
  notifies :download, "machine_file[/var/log/example]", :immediately
end

machine_file "/var/log/example" do
    local_path "/tmp/example"
    machine bootstrap_machine
    action :nothing
    not_if { one.get_resource('vm', { :name => bootstrap_machine }).nil? } 
end
```

Thinking maybe optional machine_spec attribute should be introduced to machine resources which would be used to lookup the machine_spec instead of using resource name. If not specified then you revert back to using resource name to lookup machine_spec.

**Comment by tyler-ball** [comment]
Hey @poliva83 sorry for the delayed feedback.  I can definitely repro this locally and it seems a bug to me.  My `machine_file` is notified from the first resource `:destroy` action instead of the `:converge` action.

I'm still digging into it to understand why this is happening, and then I'll work on a fix.  I'll also experiment with a workaround in the mean time (if I can find one).

**Comment by poliva83** [comment]
@tyler-ball I believe it happens because of how notifications actually work in chef. They are shared because both resource blocks using same resource name. Unfortunately I must keep that name attribute the same or otherwise the machine blocks would search for different machine_specs.

Thanks for looking into this.

**Comment by tyler-ball** [comment]
@poliva83 You're exactly right - its because the notification framework in Chef only looks at the resource name, and not the notifying action - https://github.com/chef/chef/blob/master/lib/chef/run_context.rb#L176-L183.  Any update to the notifying resource will cause the notification to trigger.  I think this is a little more obvious through the [subscribes](https://docs.chef.io/resource_common.html#subscribes-syntax) syntax where you only specify the _resource name_ to listen to, not what action to trigger on.

Because this is a core Chef feature I really don't want to introduce a hack to work around it.  I'm going to page @jkeiser to see if he knows of a workaround or pattern to deal with this.

**Comment by jkeiser** [comment]
Yep, `name` is the actual thing used for the resource name, and since we're using that attribute for the machine name as well, there's no workaround at present. I think it's totally reasonable for us to introduce a `machine_name` attribute for this, since this is the Chef model for notifications.

@poliva83 the only way I can see to solve your problem in the current Chef world is to dash off a quick resource:

```ruby
# libraries/logged_machine.rb
class DownloadingMachine < Chef::Resource::Machine
  resource_name :downloading_machine
  attribute :download_files, kind_of: Hash
end

class DownloadingMachineProvider < Chef::Provider::Machine
  provides :downloading_machine
  action :converge do
    super()
    if resource_updated?
      new_resource.download_files.each do |remote, local|
        machine_file remote do
          local_path local
          machine new_resource.name
          action :download
        end
      end
  end
end
```

Then you can use it like

```ruby
downloading_machine bootstrap_machine do
  tag ''
  machine_options :bootstrap_options => { :template => bootstrap_machine_template }
  chef_environment "#{bootstrap_environment}"
  run_list [ 'recipe[bb_apt::default]' ]
  download_files '/var/log/example' => '/tmp/example'
  action :converge
  notifies :download, "machine_file[/var/log/example]", :immediately
end
```

**Comment by jkeiser** [comment]
(And thus don't need the `machine_file`)

**Comment by tyler-ball** [comment]
Oh doy, there is already a [`files`](https://github.com/chef/chef-provisioning/blob/master/lib/chef/resource/machine.rb#L62-L67) attribute on the machine_resource that will _upload_ files to the machine.  Would that work @poliva83 ?

**Comment by jkeiser** [comment]
@poliva83 another thing you can do is get rid of the `:immediately`. Then the notification won't fire until the end of the run, and things will be fine. Depends if you're doing anything with that file, though.

**Comment by poliva83** [comment]
@jkeiser @tyler-ball The example above I provided was not the actual real business case for where I wanted to use `:notifies` on machine resource. I only provided it to illustrate the problem.

The real problem involved triggering `:snapshot` action on a custom `one_image` resource we created in our chef-provisioning driver. The recipe calls :destroy machine at start because it wants to :destroy any existing image snapshot but can't if original machine (that already converged last time I ran recipe) is running and still using the image.

My workaround with adding guards on the resource being notified is currently ok but I would be :+1: to idea of introducing a `machine_name` attribute. Main reason I created this issue was because the behaviour likly isn't expected to chef newcomers.

**Comment by tyler-ball** [comment]
@poliva83 I agree - the behavior is strange, but it has been that way in core chef for a while.  https://docs.chef.io/resource_common.html#notifications states that notifications trigger when the "resource’s state changes" so I think the docs are in line with the behavior.  But it says something that both of us expected it to behave differently :)

What would you think of having the `:snapshot` action on the `one_image` resource run
```ruby
machine "named_machine" do
  action :destroy
end
```

At the very beginning of the snapshot?  This would ensure that the machine isn't running.  This would look much nicer than having to decorate everything with guards.

I'm also a fan of (per @jkeiser's [suggestion](https://github.com/chef/chef-provisioning/issues/421#issuecomment-135830130)) writing custom resources to handle the 'do a series of steps' processes that Chef doesn't handle very well out of the box with its convergent behavior.

**Comment by poliva83** [comment]
@tyler-ball In my case I am taking OS disk snapshot of a running VM. 

`notifies :snapshot, "one_image[snap_image]", :immediately` is  on the machine block using :converge action since I only want to take snapshot after the node has converged with chef. So putting a destroy machine inside one_image's provider for :snapshot action doesn't make sense.

It just seems backwards that `machine` resources have a `name` attribute which doesn't allow you to separate resource name from attribute used to grab the machine spec. 

Would you expect the following recipe to call ruby block twice and sleep total of 60 seconds?
```
file 'delete_my_file' do
  path '/var/www/customers/public_html/index.php'
  action :delete
end

file 'create_my_file' do
  content '<html>This is a placeholder for the home page.</html>'
  mode '0755'
  path '/var/www/customers/public_html/index.php'
  owner 'web_admin'
  group 'web_admin'
  action :converge
  notifies :run, "ruby_block[sleep_30_seconds]", :immediately
end

ruby_block "sleep_30_seconds" do
  block do
    sleep 30
  end
  action :nothing
end
```
 I'd rather work towards making the `machine` resource behave like other common chef resources.

**Comment by tyler-ball** [comment]
The first resource is stored in the notification collection with the key `file[delete_my_file]` and the second resource is stored with the key `file[create_my_file]`.  So no, it would only sleep 30 seconds because the notifies are tied to different keys in the notification collection.

The issue with your `machine` [example](https://github.com/chef/chef-provisioning/issues/421#issue-102404387) is that you're specifying the `name` attribute the same for both machines (which, as you say, you must do so the machine_spec can be looked up).  So both of those resources have the same key in the notification collection - `"machine[#{bootstrap_machine}]"`.  You could write your recipe as this example and it would have the same problem:

```ruby
machine bootstrap_machine do
  action :destroy
end

machine bootstrap_machine do
  tag ''
  machine_options :bootstrap_options => { :template => bootstrap_machine_template }
  chef_environment "#{bootstrap_environment}"
  run_list [ 'recipe[bb_apt::default]' ]
  action :converge
  notifies :download, "machine_file[/var/log/example]", :immediately
end

machine_file "/var/log/example" do
    local_path "/tmp/example"
    machine bootstrap_machine
    action :nothing
    not_if { one.get_resource('vm', { :name => bootstrap_machine }).nil? } 
end
```

Thats why the notifications trigger off both resources.

So these machine resources _are_ behaving like other common chef resources.  Lets look at your file example again, slightly modified.  Basically, we're going to leave out the `path` attribute and put it in the `name` field (which path defaults to if it isn't supplied):

```ruby
file '/var/www/customers/public_html/index.php' do
  action :delete
end

file '/var/www/customers/public_html/index.php' do
  content '<html>This is a placeholder for the home page.</html>'
  mode '0755'
  owner 'web_admin'
  group 'web_admin'
  action :converge
  notifies :run, "ruby_block[sleep_30_seconds]", :immediately
end

ruby_block "sleep_30_seconds" do
  block do
    sleep 30
  end
  action :nothing
end
```

In this case I _would_ expect to see 60 seconds of sleeping.  

----------

I agree with you that there should be a surrogate key to link a resource to a machine spec.  Like `file` has a `path` attribute that defaults to the `name`, `machine` resource should have a `machine_spec_identifier` that defaults to the `name` but can be overridden.  The `name` should not be _required_ to be the primary key.

**Comment by tyler-ball** [comment]
I realize now after reading your last comment and my last comment that you probably already knew everything that I tried to explain - apologies for going over something that I think you already understand :)

I'm going to leave this issue open and change the title to reflect my current understanding.  Yes, you should be able to have 2 different resources which trigger different notifications, with both tying into the same `machine_spec`.

We could do this by introducing a `machine_spec_identifier` attribute which defaults to the resource name but can be overridden if users need distinct notifications from each resource.

I'm not sure when this will get prioritized (since you have a workaround for now) but I'll leave it as a feature request so we can triage it.

**Comment by poliva83** [comment]
@tyler-ball Apologies for not being clear in previous message. I did understand everything you were explaining :)
 
Thanks again for moving this issue forward.

**Comment by Saburesh07** [comment]
This issue or pull request is being automatically closed because the repository is being archived as part of the [Repository Standardization Initiative](https://github.com/chef-boneyard/oss-repo-standardization-2025).
